File: server/src/main/java/com/krillsson/sysapi/core/monitoring/monitors/CpuTemperatureMonitor.java
package com.krillsson.sysapi.core.monitoring.monitors;

import com.krillsson.sysapi.core.domain.system.SystemLoad;
import com.krillsson.sysapi.core.monitoring.Monitor;

import java.time.Duration;

public class CpuTemperatureMonitor extends Monitor {
    public CpuTemperatureMonitor(String id, Duration inertia, double threshold) {
        super(id, inertia, threshold);
    }

    @Override
    protected double value(SystemLoad systemLoad) {
        return systemLoad.getCpuLoad().getCpuHealth().getTemperatures().stream().findFirst().orElse(-1.0);
    }

    @Override
    protected boolean isAboveThreshold(double value) {
        return value > threshold();
    }

    @Override
    protected MonitorType type() {
        return MonitorType.CPU_TEMP;
    }
}

File: server/src/main/java/com/krillsson/sysapi/core/monitoring/monitors/CpuTemperatureMonitor.kt
package com.krillsson.sysapi.core.monitoring.monitors

import com.krillsson.sysapi.core.domain.monitor.MonitorConfig
import com.krillsson.sysapi.core.domain.monitor.MonitoredValue
import com.krillsson.sysapi.core.monitoring.Monitor
import com.krillsson.sysapi.core.monitoring.MonitorMetricQueryEvent
import java.util.*

class CpuTemperatureMonitor(override val id: UUID, override val config: MonitorConfig<MonitoredValue.NumericalValue>) : Monitor<MonitoredValue.NumericalValue>() {

    companion object {
        val selector: NumericalValueSelector = { load, _ ->
            MonitoredValue.NumericalValue(load.cpuLoad.cpuHealth.temperatures.firstOrNull()?.toLong() ?: 0)
        }
    }

    override val type: Type = Type.CPU_TEMP

    override fun selectValue(event: MonitorMetricQueryEvent): MonitoredValue.NumericalValue? =
        selector(event.load, null)

    override fun isPastThreshold(value: MonitoredValue.NumericalValue): Boolean {
        return value > config.threshold
    }
}